flink-libraries/flink-table/src/main/scala/org/apache/flink/table/api/table.scala

```diff
@@ -999,6 +1000,195 @@ class Table(
+      case _: TableFunctionCall | _: AggFunctionCall | _: Aggregation =>
+        throw new ValidationException("Only ScalarFunction can be used in map.")
+      case _ =>
+    }
+
+    val expandedFields = expandProjectList(Seq(Flattening(mapFunction)), logicalPlan, tableEnv)
+    new Table(tableEnv,
+      Project(expandedFields.map(UnresolvedAlias), logicalPlan).validate(tableEnv))
+  }
+
+  /**
+    * Performs a map operation with an user-defined scalar function or built-in scalar function.
+    * The output will be flattened if the output type is a composite type.
+    *
+    * Scala Example:
+    * {{{
+    *   class MyMapFunction extends ScalarFunction {
+    *     def eval(str: String): Row = {
+    *       Row.of(str, str)
+    *     }
+    *
+    *     def getResultType(signature: Array[Class[_]]): TypeInformation[_] =
+    *       Types.ROW(Types.STRING, Types.STRING)
+    *   }
+    *
+    *   val func = new MyMapFunction()
+    *   table.map(func('c)).as('a, 'b)
+    * }}}
+    *
+    * Java Example:
+    * {{{
+    *   class MyMapFunction extends ScalarFunction {
+    *     public Row eval(String str) {
+    *       return Row.of(str, str);
+    *     }
+    *
+    *     public TypeInformation getResultType(Class[] signature) {
+    *       return Types.ROW(Types.STRING, Types.STRING);
+    *     }
+    *   }
+    *
+    *   ScalarFunction func = new MyMapFunction();
+    *   tableEnv.registerFunction("func", func);
+    *   table.map("func(c)").as("a, b");
+    * }}}
+    */
+  def map(mapFunction: String): Table = {
+    map(ExpressionParser.parseExpression(mapFunction))
+  }
+
+  /**
+    * Performs a flatMap operation with an user-defined table function.
+    *
+    * Scala Example:
+    * {{{
+    *   class MyFlatMapFunction extends TableFunction[Row] {
+    *     def eval(str : String) {
+    *       if (str.contains("#")) {
+    *         val splits = str.split("#")
+    *         collect(Row.of(splits(0), splits(1)))
+    *       }
+    *     }
+    *
+    *     def getResultType(signature: Array[Class[_]]): TypeInformation[_] =
+    *       Types.ROW(Types.STRING, Types.STRING)
+    *   }
+    *
+    *   val func = new MyFlatMapFunction()
+    *   table.flatMap(func('c)).as('a, 'b)
+    * }}}
+    *
+    * Java Example:
+    * {{{
+    *   class MyFlatMapFunction extends TableFunction<Row> {
+    *     public void eval(String str) {
+    *       if (str.contains("#")) {
+    *         String[] splits = str.split("#");
+    *         collect(Row.of(splits[0], splits[1]));
+    *       }
+    *     }
+    *
+    *     public TypeInformation getResultType(Class[] signature) {
+    *       return Types.ROW(Types.STRING, Types.STRING);
+    *     }
+    *   }
+    *
+    *   TableFunction func = new MyFlatMapFunction();
+    *   tableEnv.registerFunction("func", func);
+    *   table.flatMap("func(c)").as("a, b");
+    * }}}
+    */
+  def flatMap(flatMapFunction: Expression): Table = {
```

Comment:
`flatMapFunction`->`tableFunction`? duo to we not introduce the new flatmap function interface.